contracts/adapters/AutoRepay.sol

```diff
@@ -0,0 +1,313 @@
+    address debtATokenAddress = _getReserveData(repayParams.debtAsset).aTokenAddress;
+    if (repayParams.collateralAsset != repayParams.debtAsset) {
+      uint256 amounts0 = _getAmountsIn(
+        repayParams.useATokenAsFrom ? collateralATokenAddress : repayParams.collateralAsset,
+        repayParams.useATokenAsTo ? debtATokenAddress : repayParams.debtAsset,
+        repayParams.debtRepayAmount.add(premium),
+        repayParams.useEthPath
+      )[0];
+      require(amounts0 <= repayParams.collateralAmount, 'slippage too high');
+      uint256 feeAmount = amounts0.mul(FEE).div(FEE_DECIMALS);
+
+      if (repayParams.useATokenAsFrom) {
+        // Transfer aTokens from user to contract address
+        _transferATokenToContractAddress(
+          collateralATokenAddress,
+          repayParams.user,
+          amounts0.add(feeAmount),
+          permitSignature
+        );
+        IERC20(collateralATokenAddress).safeTransfer(caller, feeAmount);
+      } else {
+        // Pull aTokens from user
+        _transferATokenToContractAddress(
+          collateralATokenAddress,
+          repayParams.user,
+          amounts0.add(feeAmount),
+          permitSignature
+        );
+        IERC20(collateralATokenAddress).safeTransfer(caller, feeAmount);
+        LENDING_POOL.withdraw(repayParams.collateralAsset, amounts0, address(this));
+      }
+
+      // Swap collateral asset to the debt asset
+      _swapTokensForExactTokens(
+        repayParams.collateralAsset,
+        repayParams.debtAsset,
+        repayParams.useATokenAsFrom ? collateralATokenAddress : repayParams.collateralAsset,
+        repayParams.useATokenAsTo ? debtATokenAddress : repayParams.debtAsset,
+        amounts0,
+        repayParams.debtRepayAmount.add(premium),
+        repayParams.useEthPath
+      );
+
+      if (repayParams.useATokenAsTo) {
+        // withdraw debt AToken
+        LENDING_POOL.withdraw(
+          repayParams.debtAsset,
+          IERC20(debtATokenAddress).balanceOf(address(this)),
+          address(this)
+        );
+      }
+    } else {
+      uint256 feeAmount = repayParams.debtRepayAmount.mul(FEE).div(FEE_DECIMALS);
+      uint256 aTokenTransferAmount = repayParams.debtRepayAmount.add(premium).add(feeAmount);
+      _transferATokenToContractAddress(
+        collateralATokenAddress,
+        repayParams.user,
+        aTokenTransferAmount,
+        permitSignature
+      );
+      uint256 contractBalanceBefore = IERC20(collateralATokenAddress).balanceOf(address(this));
+      LENDING_POOL.withdraw(
+        repayParams.collateralAsset,
+        repayParams.debtRepayAmount.add(premium),
+        address(this)
+      );
+      uint256 contractBalanceAfter = IERC20(collateralATokenAddress).balanceOf(address(this));
+      IERC20(collateralATokenAddress).safeTransfer(
+        caller,
+        aTokenTransferAmount - (contractBalanceBefore - contractBalanceAfter)
```

Comment:
Probably not needed, but maybe better to use safeMath like in other places

Reply:
Why not withdraw everything that is left?